site/content/blog/identify-ghost-diff-during-kc-reconcile.md

```diff
@@ -0,0 +1,276 @@
+---
+title: "Identify ghost diff during kapp Controller reconciliation"
+slug: identify-ghost-diff-during-kapp-controller-reconciliation
+date: 2022-04-21
+author: Rohit Aggarwal
+excerpt: "Identify ghost diff during kapp Controller reconciliation "
+image: /img/logo.svg
+tags: ['carvel', 'kapp-controller', 'gitops', 'diffs', 'diff']
+---
+
+# Identify ghost diff during kapp Controller reconciliation
+
+[kapp controller](https://carvel.dev/kapp-controller/), a Package manager, is compatible with Gitops philosophy. It is continuously ensuring that the cluster is or converging towards the desired state. It does so by running the reconciliation loop after every `syncPeriod` duration. In each reconciliation cycle, it monitors the current state of the resources on the cluster and tries to bring it to the desired state if there is any mismatch. It does so with the help of [kapp](https://carvel.dev/kapp/). 
+
+kapp, another carvel tool, tries to do a diff of the current live state of the resources with the desired state. Sometimes, even though there is no change in the desired state by the user explicitly, kapp still thinks some resources have changed and tries to redeploy them as part of the reconciliation. We call them `ghost` diffs. `Ghost` diffs can appear due to different reasons: controller can change some configuration of the resource, `HorizontalPodAutoscaler` influencing the no. of replicas for a deployment, etc. Everytime a `ghost` diff is detected, a new configMap gets created. kapp store its exit status and summary of operations(e.g. no. of updated/deleted/created resources) done in this configmap. If you have many packages creating `ghost` diffs in the Kubernetes(K8s) cluster, you will end up having large no. of unnecessary configMaps which can slow the K8s API server response time.
```

Comment:
> by the user explicitly

remove explicitly.

> We call them `ghost` diffs.

that sounds like it's a problem, but it's not. not a fan of "ghost".

Reply:
@cppforlife What do you think about `phantom` instead of `ghost`?

Reply:
same connotation. the point is that those diffs are not "hidden" -- they arise because something/someone is changing actualy state as compared to desired state. this docs page might be good to link to btw: https://carvel.dev/kapp/docs/v0.46.0/merge-method/. we probably want to add a sentence about why it happens and how to address them (via rebase rules).

Reply:
Adopted